spec/spec.md

```diff
@@ -0,0 +1,109 @@
+interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+### Terminology
+
+This specification defines the following terms:
+
+#### SDK
+
+An "SDK" or "Software Development Kit" is defined as one or more libraries that offers a set of public interfaces, methods, classes and attributes offering a programming model for Dapr native to a given programming language.
+
+#### Domain Object
+
+A "domain object" is defined as an instance of a class from the application's code and not part of the SDK.
+
+## Versioning
+
+### Spec Versioning
+
+This spec follows [semantic versioning 2.0](https://semver.org/spec/v2.0.0.html). This spec also directly matches the Dapr runtime API version for major and minor. It means that for each minor version release of Dapr runtime, there will be a new release of the spec with the same version. The patch version of the spec and runtime increase independently based on fixes on each. For example: The spec version 1.9 is based on the runtime version 1.9, regardless of the patch version of each.
+
+The spec will be worked on in the main branch and include the `-alpha` suffix until it reached a stable version. Once the first stable version of the spec is reached, it will be under the release-x.y branch and tagged with vX.Y.0, such as the runtime release process.
+
+This allows SDK developers to claim compliance towards a fixed spec and runtime version instead of a moving target.
+
+### SDK Versioning
+
+An SDK MUST start as one repository in the [dapr-sandbox organization](https://github.com/dapr-sandbox) at version 0.1, following the process to onboard a new Dapr sandbox project. Before an SDK can graduate to the [dapr organization](https://github.com/dapr), it MUST comply to a stable version of this spec, in addition to the graduation criteria for Dapr sandbox projects.
+
+## Client Object
+
+The SDK MUST offer a way to instantiate a client object without requiring any parameter, using the `DAPR_HTTP_PORT` or `DAPR_GRPC_PORT` environment variables injected by the Dapr sidecar injector or Dapr CLI - defaulting to `3500` or `50001`, correspondingly, if environment variable is not set.
+
+The client object MUST connect to the Dapr sidecar based on the `DAPR_SIDECAR_IP` environment variable, defaulting to `127.0.0.1` if not set.
+
+The client object MUST abstract the protocol used to communicate to the Dapr sidecar and not offer gRPC or HTTP specific input or output.
+
+The client object MIGHT offer both gRPC or HTTP implementations.
+
+The client object SHOULD NOT implement any Actor functionality since that is handled as a separate construct.
+
+The client object MUST implement all required methods for all [core building blocks](#core-building-blocks).
+
+The client object MIGHT provide overloaded methods with reduced number of parameters for simpler method invocations.
+
+The client object SHOULD NOT implement any of the methods for the [preview APIs](#alpha-apis)
```

Comment:
I see what we're suggesting for the alpha APIs below, but I don't think this is a good solution. See my comment there but I think we should do this differently if we're able to.

Reply:
I was expecting this to trigger a discussion since each SDK is doing something different here.